### PR TITLE
Flashrom refactoring

### PR DESCRIPTION
**Pull request by santhosh-kt**
#### Why I did it
To build flashrom properly with dependency tracking.

#### How I did it
Moved flashrom code from platform/broadcom/sonic-platform-modules-dell/tools directory to src/flashrom directory.
At the end, flashrom_0.9.7_amd64.deb package is build which will be installed in the devices.

#### How to verify it
Attached logs as reference.
[New Logs.txt](https://github.com/Azure/sonic-buildimage/files/6059115/New.Logs.txt)

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012

#### Description for the changelog

On branch flashrom-refactoring
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        modified:   ../../build_debian.sh
        modified:   ../../files/build_templates/sonic_debian_extension.j2
        modified:   ../../platform/broadcom/platform-modules-dell.mk
        modified:   ../../platform/broadcom/sonic-platform-modules-dell/debian/platform-modules-s6100.install
        deleted:    ../../platform/broadcom/sonic-platform-modules-dell/tools/flashrom.sh
        new file:   ../../rules/flashrom.dep
        new file:   ../../rules/flashrom.mk
        modified:   ../../slave.mk
        new file:   Makefile
        renamed:    ../../platform/broadcom/sonic-platform-modules-dell/tools/0002-Flashrom-support-for-Intel-Rangeley-and-Denverton-CP.patch -> patch/0002-Flashrom-support-for-Intel-Rangeley-and-Denverton-CP.patch
        new file:   patch/0003-Debian-package-base.patch
        new file:   patch/0004-Flashrom-remove-tmpl.patch
        new file:   patch/series

### CODE REVIEW

File: rules/flashrom.dep [hunk 1]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (jleveque) [inline rules/flashrom.dep:1]: Remove leading blank line

File: rules/flashrom.dep [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+SPATH       := $($(FLASHROM)_SRC_PATH)
+DEP_FILES   := $(SONIC_COMMON_FILES_LIST) rules/flashrom.mk rules/flashrom.dep
+DEP_FILES   += $(SONIC_COMMON_BASE_FILES_LIST)
+DEP_FILES   += $(shell git ls-files $(SPATH))
+
+$(FLASHROM)_CACHE_MODE  := GIT_CONTENT_SHA
+$(FLASHROM)_DEP_FLAGS   := $(SONIC_COMMON_FLAGS_LIST)
+$(FLASHROM)_DEP_FILES   := $(DEP_FILES)
+
```
Comment (jleveque) [inline rules/flashrom.dep:10]: Remove trailing blank line

File: src/flashrom/Makefile [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	stg init
+	stg import -s ../patch/series
+
+	# Build package
+	dpkg-buildpackage -rfakeroot -b -us -uc -j$(SONIC_CONFIG_MAKE_JOBS)
+	popd
+
+	mv $* $(DEST)/
+
+
```
Comment (jleveque) [inline src/flashrom/Makefile:27]: Remove trailing blank lines

**Review by jleveque — CHANGES_REQUESTED** [review]

**Review by jleveque — APPROVED** [review]
Looks good to me. @lguohan to review also.

**Comment by dflynn-Nokia** [comment]
Hi Santhosh,

It seems this change has broken the marvell-armhf platform build.
See output below.
Can you take a look?

Here's the steps to build for marvell-armhf

make init
make configure PLATFORM=marvell-armhf PLATFORM_ARCH=armhf
make target/debs/buster/flashrom_0.9.7_armhf.deb

Thanks

[ FAIL LOG START ] [ target/debs/buster/flashrom_0.9.7_armhf.deb ]
[ REASON ] :      target/debs/buster/flashrom_0.9.7_armhf.deb does not exist  
[ FLAGS  FILE    ] : [] 
[ FLAGS  DEPENDS ] : [] 
[ FLAGS  DIFF    ] : [] 
make[1]: Entering directory '/sonic/src/flashrom'
rm -rf ./flashrom-0.9.7
# Obtain flashrom
git clone https://github.com/flashrom/flashrom.git flashrom-0.9.7
pushd ./flashrom-0.9.7
# Check out tag: tags/0.9.7
git checkout -b flashrom-src tags/0.9.7
# Apply patch series
stg init
stg import -s ../patch/series
# Build package
dpkg-buildpackage -rfakeroot -b -us -uc -j8
popd
mv flashrom_0.9.7_armhf.deb /sonic/target/debs/buster/
Cloning into 'flashrom-0.9.7'...
/sonic/src/flashrom/flashrom-0.9.7 /sonic/src/flashrom
Switched to a new branch 'flashrom-src'
dpkg-buildpackage: info: source package flashrom
dpkg-buildpackage: info: source version 0.9.7
dpkg-buildpackage: info: source distribution unstable
dpkg-buildpackage: info: source changed by Dell Team <support@dell.com>
 dpkg-source --before-build .
dpkg-buildpackage: info: host architecture armhf
 fakeroot debian/rules clean
make[2]: Entering directory '/sonic/src/flashrom/flashrom-0.9.7'
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
dh clean
make[2]: Leaving directory '/sonic/src/flashrom/flashrom-0.9.7'
 debian/rules build
make[2]: Entering directory '/sonic/src/flashrom/flashrom-0.9.7'
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
dh build
make[2]: Leaving directory '/sonic/src/flashrom/flashrom-0.9.7'
 fakeroot debian/rules binary
make[2]: Entering directory '/sonic/src/flashrom/flashrom-0.9.7'
make[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
dh binary
make[2]: Leaving directory '/sonic/src/flashrom/flashrom-0.9.7'
 dpkg-genbuildinfo --build=binary
dpkg-genbuildinfo: error: binary build with no binary artifacts found; .buildinfo is meaningless
dpkg-buildpackage: error: dpkg-genbuildinfo --build=binary subprocess returned exit status 25
make[1]: *** [Makefile:9: /sonic/target/debs/buster/flashrom_0.9.7_armhf.deb] Error 25
make[1]: Leaving directory '/sonic/src/flashrom'
[  FAIL LOG END  ] [ target/debs/buster/flashrom_0.9.7_armhf.deb ]
make: *** [slave.mk:418: target/debs/buster/flashrom_0.9.7_armhf.deb] Error 1
DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. pip 21.0 will drop support for Python 2.7 in January 2021. More details about Python 2 support in pip can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support pip 21.0 will remove support for this functionality.
Makefile.work:280: recipe for target 'target/debs/buster/flashrom_0.9.7_armhf.deb' failed
make[1]: *** [target/debs/buster/flashrom_0.9.7_armhf.deb] Error 2
make[1]: Leaving directory '/home/dflynn/data/git-repos/github-sonic-buildimage'
Makefile:7: recipe for target 'target/debs/buster/flashrom_0.9.7_armhf.deb' failed

**Pull request merged**